File: 1090-largest-values-from-labels/1090-largest-values-from-labels.py
from collections import defaultdict
class Solution:
    def largestValsFromLabels(self, values: List[int], labels: List[int], numWanted: int, useLimit: int) -> int:
        
        #Sorting in decending order based on "value"
        items = sorted(zip(values, labels), key=lambda x: -x[0])
        print(items)

        my_dict = dict()
        res = 0

        for value, label in items:
            
            if numWanted == 0:
                break

            if my_dict.get(label, 0) < useLimit:
                res += value
                my_dict[label] = my_dict.get(label, 0) + 1
                numWanted -= 1
    

        return res
                

File: 1090-largest-values-from-labels/1090-largest-values-from-labels.cpp
class Solution {
public:
        int largestValsFromLabels(std::vector<int>& values, std::vector<int>& labels, int num_wanted, int use_limit)
{
    auto zip = [](const std::vector<int>& a, const std::vector<int>& b)
    {
        std::vector<std::pair<int, int>> result;
        size_t size = std::min(a.size(), b.size());
        result.reserve(size);
        for (size_t i = 0; i < size; ++i) {
            result.emplace_back(a[i], b[i]);
        }
        return result;
    };

    auto pairs = zip(values, labels);

    // Sort by descending value
    std::sort(pairs.begin(), pairs.end(), [](const auto& a, const auto& b) {
        return a.first > b.first;
    });

    std::unordered_map<int, int> label_count;
    int total = 0;
    int picked = 0;

    for (const auto& [value, label] : pairs) {
        if (picked == num_wanted) break;
        if (label_count[label] < use_limit) {
            total += value;
            label_count[label]++;
            picked++;
        }
    }

    return total;
}
};